src/javatests/com/google/cloud/bigquery/dwhassessment/hooks/logger/EventLoggerTest.java

```diff
@@ -0,0 +1,208 @@
+                .set("Timestamp", QUERY_COMPLETED_TIME)
+                .set("RequestUser", "test_user")
+                .set("User", System.getProperty("user.name"))
+                .set("OperationId", "test_op_id")
+                .set("Status", "SUCCESS")
+                .set("PerfObject", "{}")
+                .set("HookVersion", "1.0")
+                .build());
+  }
+
+  @Test
+  public void onFailureHook_success() throws Exception {
+    String queryText = "SELECT * FROM employees";
+    String queryId = "hive_query_id_999";
+    QueryPlan queryPlan = createQueryPlan(queryText, queryId);
+    HookContext context = createContext(queryPlan);
+    context.setHookType(HookType.ON_FAILURE_HOOK);
+
+    // Act
+    logger.handle(context);
+    logger.shutdown();
+
+    // Assert
+    List<GenericRecord> records = readOutputRecords(conf, tmpFolder);
+    assertThat(records)
+        .containsExactly(
+            new GenericRecordBuilder(QUERY_EVENT_SCHEMA)
+                .set("QueryId", queryId)
+                .set("EventType", "QUERY_COMPLETED")
+                .set("Timestamp", QUERY_COMPLETED_TIME)
+                .set("RequestUser", "test_user")
+                .set("User", System.getProperty("user.name"))
+                .set("OperationId", "test_op_id")
+                .set("Status", "FAIL")
+                .set("PerfObject", "{}")
+                .set("HookVersion", "1.0")
+                .build());
+  }
+
+  private HookContext createContext(QueryPlan queryPlan) throws Exception {
+    PerfLogger perfLogger = PerfLogger.getPerfLogger(conf, true);
+    return new HookContext(
+        queryPlan,
+        queryState,
+        null,
+        "test_user",
+        "192.168.10.10",
+        "hive_addr",
+        "test_op_id",
+        "test_session_id",
+        "test_thread_id",
+        true,
+        perfLogger);
+  }
+
+  private QueryPlan createQueryPlan(String queryText, String queryId) throws Exception {
+    BaseSemanticAnalyzer sem = new DDLSemanticAnalyzer(queryState, hiveMock);
+
+    return new QueryPlan(queryText, sem, 1234L, queryId, HiveOperation.QUERY, null);
+  }
+
+  public static List<GenericRecord> readOutputRecords(HiveConf conf, String tmpFolder)
+      throws IOException {
+    Path path = new Path(tmpFolder);
+    FileSystem fs = path.getFileSystem(conf);
+
+    ImmutableList<FileStatus> directories = ImmutableList.copyOf(fs.listStatus(path));
+    assertThat(directories).hasSize(1);
+    ImmutableList<FileStatus> files =
+        ImmutableList.copyOf(fs.listStatus(directories.get(0).getPath()));
+    assertThat(files).hasSize(1);
+
+    FSDataInputStream inputStream = fs.open(files.get(0).getPath());
+
+    DatumReader<GenericRecord> reader = new GenericDatumReader<>(QUERY_EVENT_SCHEMA);
+
+    try (DataFileStream<GenericRecord> dataFileReader = new DataFileStream<>(inputStream, reader)) {
+      ArrayList<GenericRecord> records = new ArrayList<>();
```

Comment:
I would suggest to change the type of the variable to `List<GenericRecord>` (an interface).

Reply:
Done